api/chat.go

```diff
@@ -45,61 +45,124 @@ func configGinChatRouter(router *gin.RouterGroup) {
 		if tumLiveContext.User == nil {
 			return
 		}
-		var chat ChatReq
-		if err := json.Unmarshal(msg, &chat); err != nil {
-			return
-		}
-		if !tumLiveContext.Course.ChatEnabled {
-			return
-		}
-		uname := tumLiveContext.User.Name
-		if chat.Anonymous && tumLiveContext.Course.AnonymousChatEnabled {
-			uname = "Anonymous"
-		}
-		replyTo := sql.NullInt64{}
-		if chat.ReplyTo == 0 {
-			replyTo.Int64 = 0
-			replyTo.Valid = false
-		} else {
-			replyTo.Int64 = chat.ReplyTo
-			replyTo.Valid = true
-		}
-		chatForDb := model.Chat{
-			UserID:   strconv.Itoa(int(tumLiveContext.User.ID)),
-			UserName: uname,
-			Message:  chat.Msg,
-			StreamID: tumLiveContext.Stream.ID,
-			Admin:    tumLiveContext.User.ID == tumLiveContext.Course.UserID,
-			ReplyTo:  replyTo,
-		}
-		err := dao.AddMessage(&chatForDb)
+		var req wsReq
+		err := json.Unmarshal(msg, &req)
 		if err != nil {
-			if errors.Is(err, model.ErrCooledDown) {
-				sendServerMessage("You are sending messages too fast. Please wait a bit.", TypeServerErr, s)
-			}
+			log.Warn(err)
 			return
 		}
-		if broadcast, err := json.Marshal(chatForDb); err == nil {
-			broadcastStream(tumLiveContext.Stream.ID, broadcast)
+		if req.Type == "message" {
+			handleMessage(tumLiveContext, s, msg)
+		} else if req.Type == "like" {
+			handleLike(tumLiveContext, msg)
 		}
 	})
 }
 
+func handleLike(ctx tools.TUMLiveContext, msg []byte) {
+	var req likeReq
+	err := json.Unmarshal(msg, &req)
+	if err != nil {
+		return
+	}
+	err = dao.ToggleLike(ctx.User.ID, req.Id)
+	if err != nil {
+		log.WithError(err).Error("error liking/unliking message")
+		return
+	}
+	numLikes, err := dao.GetNumLikes(req.Id)
+	if err != nil {
+		log.WithError(err).Error("error getting num of chat likes")
+		return
+	}
+	broadcast := gin.H{
+		"likes": req.Id,
+		"num":   numLikes,
+	}
+	broadcastBytes, err := json.Marshal(broadcast)
+	if err != nil {
+		log.WithError(err).Warn("Can't marshal like message")
```

Comment:
Should also be consistent to Unmarshal operations

Reply:
Yup, this should be `Error` :)